hnn_core/network.py

```diff
@@ -245,53 +246,54 @@ def pick_connection(net, src_gids=None, target_gids=None,
     valid_receptor = ['ampa', 'nmda', 'gabaa', 'gabab']
 
     # Convert receptor and loc to list
-    loc = _string_input_to_list(loc, valid_loc, 'loc')
-    receptor = _string_input_to_list(receptor, valid_receptor, 'receptor')
+    loc_list = _string_input_to_list(loc, valid_loc, 'loc')
+    receptor_list = _string_input_to_list(receptor, valid_receptor, 'receptor')
 
     # Create lookup dictionaries
-    src_dict, target_dict = dict(), dict()
-    loc_dict, receptor_dict = dict(), dict()
+    src_dict, target_dict = defaultdict(list), defaultdict(list)
+    loc_dict, receptor_dict = defaultdict(list), defaultdict(list)
     for conn_idx, conn in enumerate(net.connectivity):
         # Store connections matching each src_gid
         for src_gid in conn['src_gids']:
-            if src_gid in src_dict:
-                src_dict[src_gid].append(conn_idx)
-            else:
-                src_dict[src_gid] = [conn_idx]
+            src_dict[src_gid].append(conn_idx)
+
         # Store connections matching each target_gid
         for target_gid in conn['target_gids']:
-            if target_gid in target_dict:
-                target_dict[target_gid].append(conn_idx)
-            else:
-                target_dict[target_gid] = [conn_idx]
+            target_dict[target_gid].append(conn_idx)
+
         # Store connections matching each location
-        if conn['loc'] in loc_dict:
-            loc_dict[conn['loc']].append(conn_idx)
-        else:
-            loc_dict[conn['loc']] = [conn_idx]
+        loc_dict[conn['loc']].append(conn_idx)
+
         # Store connections matching each receptor
-        if conn['receptor'] in receptor_dict:
-            receptor_dict[conn['receptor']].append(conn_idx)
-        else:
-            receptor_dict[conn['receptor']] = [conn_idx]
+        receptor_dict[conn['receptor']].append(conn_idx)
 
     # Look up conn indices that match search terms and add to set.
     conn_set = set()
-    search_pairs = [(src_gids, src_dict), (target_gids, target_dict),
-                    (loc, loc_dict), (receptor, receptor_dict)]
-    for search_terms, search_dict in search_pairs:
-        inner_set = set()
-        # Union of indices which match inputs for single parameter
-        for term in search_terms:
-            inner_set = inner_set.union(search_dict.get(term, list()))
-        # Intersection across parameters
-        if conn_set and inner_set:
-            conn_set = conn_set.intersection(inner_set)
-            # If at any point there's no matching elements, return empty
-            if not conn_set:
+    search_pairs = [(src_gids, src_gids_checked, src_dict),
```

Comment:
I don't follow why we need to search across both `src_gids` and `src_gids_checked` (and same for `target`, `loc`, and `receptor`) ? Isn't `_check_gids()` and `_string_input_to_list()` supposed to handle this?

Reply:
Ah good question. So what I'm doing here is preserving the input arguments to the function in the first element of the tuples. The function has default values of `None` for each argument. Basically if the argument is still `None`, it won't conduct a search on that attribute (if statement at 278). That's all the first element of the tuple is used for. 

This is a necessary change since previously we were looping across all 4 attributes regardless of if it was asked for or not. This would end up with empty searches for the non-specified attributes. This would break at the new condition to break the search if there is an empty `inner_set`.

Reply:
Hmmm I now I see that `src_gids_checked` is just an empty list if `src_gids` is None. So I could just do the line 278 check as `if search_terms: `and get rid of first elements of the tuple!